### PR TITLE
feat: Integrate userData into Progresses API to reduce redundant calls

### LINKED ISSUE #2294: Include Full User Details in Task Progress Updates
### Issue Description

Currently, when progress data is fetched, we are only sending the userId in the progress object. As a result, the frontend has to make an additional API call to fetch the user data (e.g., username, profile picture). This increases the number of API calls and affects performance. To reduce this, we plan to send the full user data directly with the progress update, eliminating the need for an additional API call on the frontend.

### Expected Behavior

- The api response should include the full user data (e.g., username, profile picture) along with the progress object whenever a task's progress is fetched.
- The progress objects will contain the user data alongside the existing progress fields, eliminating the need for the frontend to make a separate API call to fetch user details.
- Ensure compatibility with the frontend, which will render the username and profile picture directly from the progress object without additional API calls.
### Current Behavior

- Ccurrently we are only sending userId in the progress object.
- The frontend has to make additional API calls to fetch user details based on the userId for rendering, which increases the number of requests and affects performance.
simplicity.
### Reproducibility

 - [ ] This issue is reproducible
 - [ ] This issue is not reproducible

### Steps to Reproduce
1. 
2. 

### Severity/Priority
 - [ ] Critical
 - [ ] High
 - [x] Medium
 - [ ] Low

### Checklist

- [x] I have read and followed the project's code of conduct.
- [x] I have searched for similar issues before creating this one.
- [x] I have provided all the necessary information to understand and reproduce the issue.
- [x] I am willing to contribute to the resolution of this issue.

### PR DESCRIPTION
**Pull request by AnujChhikara**
Date: 21 Dec 2024

Developer Name: Anuj Chhikara

---

## Issue Ticket Number

- Closes #2294 
## Description

This PR introduces a change to include the full `userData` (e.g., username, profile picture) alongside the progress data when fetching progress updates. Currently, the backend only returns the `userId`, requiring the frontend to make an additional API call to fetch user details. By including `userData` directly with the progress object, we aim to reduce the number of API calls and improve frontend performance and simplicity.
### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [x] Yes
- [ ] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

- Without dev flag
ShareX_9fydyPImyR

- With dev flag
Postman_CKua0H9mYP

</details>

## Test Coverage
<details>
<summary>Screenshot 1</summary>

image
image
image

</details>

## Additional Notes
-  [Design Doc](https://docs.google.com/document/d/12oFgfa0mMrIMdQciZxUCJrTbzDOpF_tOoLjR9NDH8yc/edit?tab=t.0#heading=h.5fv6fjv5jjyd)

### CODE REVIEW

File: models/progresses.js [hunk 1]
```diff
@@ -47,9 +48,31 @@ const createProgressDocument = async (progressData) => {
...
+  const { dev } = queryParams;
   await assertUserOrTaskExists(queryParams);
   const query = buildQueryToFetchDocs(queryParams);
   const progressDocs = await getProgressDocs(query);
+
+  if (dev) {
+    try {
+      const uniqueUserIds = [...new Set(progressDocs.map((doc) => doc.userId))];
+      const usersData = await Promise.all(uniqueUserIds.map((userId) => fetchUser({ userId })));
+
```
Comment (Achintya-Chatterjee) [inline models/progresses.js:60]: While fetching `usersData`, the use of nested `Promise.all` could cause performance issues for large datasets. We should consider batching the user IDs and using an optimized solution.

File: test/integration/progressesTasks.test.js [hunk 2]
```diff
@@ -222,6 +222,34 @@ describe("Test Progress Updates API for Tasks", function () {
         });
     });
 
+    it("Returns the progress array for the task with userData object", function (done) {
+      chai
+        .request(app)
```
Comment (Achintya-Chatterjee) [inline test/integration/progressesTasks.test.js:227]: Can we please write a test with an empty progress dataset to ensure the response structure is as expected?

File: models/progresses.js (hunk 1 above)
Comment (AnujChhikara) [reply models/progresses.js]: I have used batches approach to fetch userIds

File: test/integration/progressesTasks.test.js (hunk 2 above)
Comment (AnujChhikara) [reply test/integration/progressesTasks.test.js:227]: added please check

File: models/progresses.js [hunk 3]
```diff
@@ -47,9 +48,40 @@ const createProgressDocument = async (progressData) => {
...
+      );
+
+      const batchPromises = batches.map((batch) => Promise.all(batch.map((userId) => fetchUser({ userId }))));
+
+      const usersDataBatches = await Promise.all(batchPromises);
+
+      const usersData = usersDataBatches.flat();
+
+      const userLookupMap = usersData.reduce((lookup, { user }) => {
+        if (user) lookup[user.id] = user;
```
Comment (Achintya-Chatterjee) [inline models/progresses.js:72]: write tests for this.

Comment (AnujChhikara) [reply models/progresses.js:72]: Hello @Achintya-Chatterjee I just need some clarification regarding the tests. I can add a test to check that userData returns null if the operation fails, but I'm also wondering how to test for scenarios with a large number of unique users. With our current test setup, I would need to add 1000+ users to Firestore to test it.

If you want me to proceed with that approach, please let me know!

**Comment by vinit717** [comment]
Please fix the title as commit message is not the title

File: models/progresses.js [hunk 4]
```diff
@@ -47,9 +48,40 @@ const createProgressDocument = async (progressData) => {
...
+  if (dev) {
+    try {
+      const uniqueUserIds = [...new Set(progressDocs.map((doc) => doc.userId))];
+      const batchSize = 500;
+
+      const batches = Array.from({ length: Math.ceil(uniqueUserIds.length / batchSize) }, (_, index) =>
+        uniqueUserIds.slice(index * batchSize, index * batchSize + batchSize)
+      );
+
+      const batchPromises = batches.map((batch) => Promise.all(batch.map((userId) => fetchUser({ userId }))));
```
Comment (vinit717) [inline models/progresses.js:65]: Please refactor this and why we need this?

Comment (AnujChhikara) [reply models/progresses.js:65]: Achintya asked me to use the batches approach as Promise.all could cause performance issues for large datasets. You can check the prevoius resolved messages

Comment (AnujChhikara) [reply models/progresses.js]: resolve

File: models/progresses.js [hunk 5]
```diff
@@ -47,9 +48,14 @@ const createProgressDocument = async (progressData) => {
  * @throws {Error} If the userId or taskId is invalid or does not exist.
  **/
 const getProgressDocument = async (queryParams) => {
+  const { dev } = queryParams;
   await assertUserOrTaskExists(queryParams);
   const query = buildQueryToFetchDocs(queryParams);
   const progressDocs = await getProgressDocs(query);
+
+  if (dev) {
```
Comment (vikasosmium) [inline models/progresses.js:56]: check this also, similar case in other PRs.

File: models/progresses.js [hunk 6]
```diff
@@ -47,9 +48,14 @@ const createProgressDocument = async (progressData) => {
  * @throws {Error} If the userId or taskId is invalid or does not exist.
  **/
 const getProgressDocument = async (queryParams) => {
+  const { dev } = queryParams;
```
Comment (vikasosmium) [inline models/progresses.js:51]: check here only if it's true or not

File: models/progresses.js [hunk 7]
```diff
@@ -47,9 +48,14 @@ const createProgressDocument = async (progressData) => {
...
   await assertUserOrTaskExists(queryParams);
   const query = buildQueryToFetchDocs(queryParams);
   const progressDocs = await getProgressDocs(query);
+
+  if (dev === "true") {
+    return await addUserDetailsToProgressDocs(progressDocs);
+  }
```
Comment (vikasosmium) [inline models/progresses.js:58]: instead of checking dev here, do an early return before all this. will save you from fetching all this.

Comment (AnujChhikara) [reply models/progresses.js:58]: The reason for this structure is that we want to run all the common logic (like fetching the progress docs) regardless of the dev flag. Only if dev is true, we add the extra step of adding userdata with addUserDetailsToProgressDocs.

File: models/progresses.js [hunk 8]
```diff
@@ -77,16 +83,58 @@ const getRangeProgressData = async (queryParams) => {
...
   }
   const doc = result.docs[0];
-  return { id: doc.id, ...doc.data() };
+  const docData = doc.data();
+  if (dev === "true") {
+    const { user: userData } = await fetchUser({ userId: docData.userId });
+    return { id: doc.id, ...docData, userData };
+  }
+
+  return { id: doc.id, ...docData };
```
Comment (pankajjs) [inline models/progresses.js:102]: Why are we doing this?

Comment (AnujChhikara) [reply models/progresses.js:102]: Dev = false or if dev query param not provided he will not send the userdata field and the api return will be same as previous

Comment (pankajjs) [reply models/progresses.js:102]: Resolved

File: models/progresses.js [hunk 9]
```diff
@@ -77,16 +83,58 @@ const getRangeProgressData = async (queryParams) => {
...
+    }, {});
+
+    return progressDocs.map((doc) => {
+      const userDetails = userByIdMap[doc.userId] || null;
+      return { ...doc, userData: userDetails };
+    });
+  } catch (err) {
+    return progressDocs.map((doc) => ({ ...doc, userData: null }));
+  }
+};
```
Comment (Achintya-Chatterjee) [inline models/progresses.js:132]: please write unit test for this function

Comment (AnujChhikara) [reply models/progresses.js:132]: added please check

**Review by Achintya-Chatterjee — DISMISSED** [review]
LGTM

**Review by Achintya-Chatterjee — APPROVED** [review]

**Review by vikasosmium — APPROVED** [review]
